presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkMutableRow.java

```diff
@@ -13,48 +13,87 @@
 
-import java.io.Serializable;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.KryoSerializable;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
 
-import static java.util.Objects.requireNonNull;
+import java.io.Externalizable;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.nio.ByteBuffer;
 
 public class PrestoSparkMutableRow
-        implements Serializable
+        implements Externalizable, KryoSerializable
 {
-    private static final int INSTANCE_SIZE = Long.BYTES * 2 /* headers */
-            + Integer.BYTES /* partition */
-            + Integer.BYTES /* length */
-            + Long.BYTES /* bytes pointer */
-            + Long.BYTES * 2 /* bytes headers */
-            + Integer.BYTES /* bytes length */;
+    private int partition;
+    private ByteBuffer buffer;
 
-    private final int partition;
-    private final int length;
-    private final byte[] bytes;
+    public int getPartition()
+    {
+        return partition;
+    }
 
-    public PrestoSparkMutableRow(int partition, int length, byte[] bytes)
+    public void setPartition(int partition)
     {
         this.partition = partition;
-        this.length = length;
-        this.bytes = requireNonNull(bytes, "bytes is null");
     }
 
-    public int getPartition()
+    public ByteBuffer getBuffer()
     {
-        return partition;
+        return buffer;
+    }
+
+    public void setBuffer(ByteBuffer buffer)
+    {
+        this.buffer = buffer;
+    }
+
+    /**
+     * TODO: Transitional method. Will be removed in the next commit.
+     */
+    public int getRetainedSize()
+    {
+        return Integer.SIZE + buffer.remaining();
+    }
+
+    public PrestoSparkMaterializedRow toMaterializedRow()
+    {
+        byte[] copy = new byte[buffer.remaining()];
+        System.arraycopy(buffer.array(), buffer.arrayOffset() + buffer.position(), copy, 0, buffer.remaining());
+        return new PrestoSparkMaterializedRow(copy);
+    }
+
+    @Override
+    public void write(Kryo kryo, Output output)
+    {
+        throw serializationNotSupportedException();
+    }
+
+    @Override
+    public void read(Kryo kryo, Input input)
+    {
+        throw serializationNotSupportedException();
     }
 
-    public int getLength()
+    @Override
+    public void writeExternal(ObjectOutput output)
     {
-        return length;
+        throw serializationNotSupportedException();
     }
 
-    public byte[] getBytes()
+    @Override
+    public void readExternal(ObjectInput input)
     {
-        return bytes;
+        throw serializationNotSupportedException();
     }
 
-    public long getRetainedSize()
+    private static RuntimeException serializationNotSupportedException()
     {
-        return INSTANCE_SIZE + bytes.length;
+        // PrestoSparkUnsafeRow is expected to be serialized only during shuffle.
+        // Shuffle rows are always serialized with PrestoSparkShuffleSerializer.
```

Comment:
So essentially, although `PrestoSparkMutableRow` claims it implements `KryoSerializable` interface, it just tricks the type system. As the serialization is actually done by `PrestoSparkShuffleSerializer`